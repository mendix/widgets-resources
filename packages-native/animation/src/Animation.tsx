import { Component, ReactNode, createElement } from "react";
import { flattenStyles } from "@native-mobile-resources/util-widgets";
import { Animation as AnimationType, View, Easing } from "react-native-animatable";

import { AnimationProps } from "../typings/AnimationProps";
import { defaultAnimationStyle, AnimationStyle } from "./ui/Styles";

export type Props = AnimationProps<AnimationStyle>;
type Direction = "normal" | "reverse" | "alternate" | "alternate-reverse";

export class Animation extends Component<Props> {
    private readonly animationEndHandle = this.onAnimationEnd.bind(this);
    private readonly styles = flattenStyles(defaultAnimationStyle, this.props.style);

    render(): ReactNode {
        const { count, duration, content, easing, delay, direction } = this.props;
        const easingValue = easing.replace(/_/g, "-") as Easing;
        const directionValue = direction.replace(/_/g, "-") as Direction;
        const countValue = count === 0 ? "infinite" : count;
        this.validateProps(this.props);

        return (
            <View
                animation={this.getAnimation()}
                duration={duration}
                delay={delay}
                direction={directionValue}
                easing={easingValue}
                iterationCount={countValue}
                onAnimationEnd={this.animationEndHandle}
                style={this.styles.container}
                useNativeDriver
            >
                {content}
            </View>
        );
    }

    private validateProps(props: Props): void {
        const { afterAnimationAction, count } = props;
        if (afterAnimationAction && count === 0) {
            this.log("After animation action can not be triggered by infinite count");
        }
        const { animationType, animationIn, animationOut, animationAttention } = this.props;
        if (animationType === "in" && animationIn === "none") {
            this.log("No 'Entry animation' is selected for animation type 'Entry'");
        }
        if (animationType === "attention" && animationAttention === "none") {
            this.log("No 'Attention animation' is selected for animation type 'Attention'");
        }
        if (animationType === "out" && animationOut === "none") {
            this.log("No 'Exit animation' is selected for animation type 'Exit'");
        }
    }

    private log(message: string): void {
        // eslint-disable-next-line no-console
        console.warn(message);
    }

    private getAnimation(): AnimationType | undefined {
        const { animationType, animationIn, animationOut, animationAttention, condition } = this.props;
        const animation =
            animationType === "in" ? animationIn : animationType === "out" ? animationOut : animationAttention;

        if (!condition || (condition.status === "available" && condition.value === true)) {
            return animation === "none" ? undefined : animation;
        }
        return undefined;
    }

    private onAnimationEnd(): void {
        if (this.props.afterAnimationAction && this.props.afterAnimationAction.canExecute) {
            this.props.afterAnimationAction.execute();
        }
    }
}
