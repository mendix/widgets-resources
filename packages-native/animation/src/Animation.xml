<?xml version="1.0" encoding="utf-8" ?>
<widget id="com.mendix.widget.native.animation.Animation" pluginWidget="true" needsEntityContext="true" offlineCapable="true"
        supportedPlatform="Native"
        xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../node_modules/mendix/custom_widget.xsd">
    <name>Animation</name>
    <description>Container for animating content</description>
    <icon/>
    <properties>
        <propertyGroup caption="General">
            <propertyGroup caption="Content">
                <property key="content" type="widgets" required="true">
                    <caption>Content</caption>
                    <description/>
                </property>
            </propertyGroup>

            <propertyGroup caption="Effect">
                <property key="animation" type="enumeration" defaultValue="fadeIn">
                    <caption>Animation</caption>
                    <description/>
                    <enumerationValues>
                        <enumerationValue key="bounce">Bounce</enumerationValue>
                        <enumerationValue key="flash">Flash</enumerationValue>
                        <enumerationValue key="jello">Jello</enumerationValue>
                        <enumerationValue key="pulse">Pulse</enumerationValue>
                        <enumerationValue key="rotate">Rotate</enumerationValue>
                        <enumerationValue key="rubberBand">Rubber Band</enumerationValue>
                        <enumerationValue key="shake">Shake</enumerationValue>
                        <enumerationValue key="swing">Swing</enumerationValue>
                        <enumerationValue key="tada">Tada</enumerationValue>
                        <enumerationValue key="wobble">Wobble</enumerationValue>
                        <enumerationValue key="bounceIn">Bounce In</enumerationValue>
                        <enumerationValue key="bounceInDown">Bounce In Down</enumerationValue>
                        <enumerationValue key="bounceInUp">Bounce In Up</enumerationValue>
                        <enumerationValue key="bounceInLeft">Bounce In Left</enumerationValue>
                        <enumerationValue key="bounceInRight">Bounce In Right</enumerationValue>
                        <enumerationValue key="bounceOut">Bounce Out</enumerationValue>
                        <enumerationValue key="bounceOutDown">Bounce Out Down</enumerationValue>
                        <enumerationValue key="bounceOutUp">Bounce Out Up</enumerationValue>
                        <enumerationValue key="bounceOutLeft">Bounce Out Left</enumerationValue>
                        <enumerationValue key="bounceOutRight">Bounce Out Right</enumerationValue>
                        <enumerationValue key="fadeIn">Fade In</enumerationValue>
                        <enumerationValue key="fadeInDown">Fade In Down</enumerationValue>
                        <enumerationValue key="fadeInDownBig">Fade In Down Big</enumerationValue>
                        <enumerationValue key="fadeInUp">Fade In Up</enumerationValue>
                        <enumerationValue key="fadeInUpBig">Fade In Up Big</enumerationValue>
                        <enumerationValue key="fadeInLeft">Fade In Left</enumerationValue>
                        <enumerationValue key="fadeInLeftBig">Fade In Left Big</enumerationValue>
                        <enumerationValue key="fadeInRight">Fade In Right</enumerationValue>
                        <enumerationValue key="fadeInRightBig">Fade In Right Big</enumerationValue>
                        <enumerationValue key="fadeOut">Fade Out</enumerationValue>
                        <enumerationValue key="fadeOutDown">Fade Out Down</enumerationValue>
                        <enumerationValue key="fadeOutDownBig">Fade Out Down Big</enumerationValue>
                        <enumerationValue key="fadeOutUp">Fade Out Up</enumerationValue>
                        <enumerationValue key="fadeOutUpBig">Fade Out Up Big</enumerationValue>
                        <enumerationValue key="fadeOutLeft">Fade Out Left</enumerationValue>
                        <enumerationValue key="fadeOutLeftBig">Fade Out Left Big</enumerationValue>
                        <enumerationValue key="fadeOutRight">Fade Out Right</enumerationValue>
                        <enumerationValue key="fadeOutRightBig">Fade Out Right Big</enumerationValue>
                        <enumerationValue key="flipInX">Flip In X</enumerationValue>
                        <enumerationValue key="flipInY">Flip In Y</enumerationValue>
                        <enumerationValue key="flipOutX">Flip Out X</enumerationValue>
                        <enumerationValue key="flipOutY">Flip Out Y</enumerationValue>
                        <enumerationValue key="lightSpeedIn">Light Speed In</enumerationValue>
                        <enumerationValue key="lightSpeedOut">Light Speed Out</enumerationValue>
                        <enumerationValue key="slideInDown">Slide In Down</enumerationValue>
                        <enumerationValue key="slideInUp">Slide In Up</enumerationValue>
                        <enumerationValue key="slideInLeft">Slide In Left</enumerationValue>
                        <enumerationValue key="slideInRight">Slide In Right</enumerationValue>
                        <enumerationValue key="slideOutDown">Slide Out Down</enumerationValue>
                        <enumerationValue key="slideOutUp">Slide Out Up</enumerationValue>
                        <enumerationValue key="slideOutLeft">Slide Out Left</enumerationValue>
                        <enumerationValue key="slideOutRight">Slide Out Right</enumerationValue>
                        <enumerationValue key="zoomIn">Zoom In</enumerationValue>
                        <enumerationValue key="zoomInDown">Zoom In Down</enumerationValue>
                        <enumerationValue key="zoomInUp">Zoom In Up</enumerationValue>
                        <enumerationValue key="zoomInLeft">Zoom In Left</enumerationValue>
                        <enumerationValue key="zoomInRight">Zoom In Right</enumerationValue>
                        <enumerationValue key="zoomOut">Zoom Out</enumerationValue>
                        <enumerationValue key="zoomOutDown">Zoom Out Down</enumerationValue>
                        <enumerationValue key="zoomOutUp">Zoom Out Up</enumerationValue>
                        <enumerationValue key="zoomOutLeft">Zoom Out Left</enumerationValue>
                        <enumerationValue key="zoomOutRight">Zoom Out Right</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="easing" type="enumeration" defaultValue="ease">
                    <caption>Easing</caption>
                    <description/>
                    <enumerationValues>
                        <enumerationValue key="linear">Linear</enumerationValue>
                        <enumerationValue key="ease">Ease</enumerationValue>
                        <enumerationValue key="ease_in">Ease in</enumerationValue>
                        <enumerationValue key="ease_out">Ease ou</enumerationValue>
                        <enumerationValue key="ease_in_out">Ease in out</enumerationValue>
                        <enumerationValue key="ease_in_cubic">Ease in cubic</enumerationValue>
                        <enumerationValue key="ease_out_cubic">Ease out cubic</enumerationValue>
                        <enumerationValue key="ease_in_out_cubic">Ease in out cubic</enumerationValue>
                        <enumerationValue key="ease_in_circ">Ease in circle</enumerationValue>
                        <enumerationValue key="ease_out_circ">Ease out circle</enumerationValue>
                        <enumerationValue key="ease_in_out_circ">Ease in out circle</enumerationValue>
                        <enumerationValue key="ease_in_expo">Ease in exponential</enumerationValue>
                        <enumerationValue key="ease_out_expo">Ease out exponential</enumerationValue>
                        <enumerationValue key="ease_in_out_expo">Ease in out exponential</enumerationValue>
                        <enumerationValue key="ease_in_quad">Ease in quardatic</enumerationValue>
                        <enumerationValue key="ease_out_quad">Ease out quardatic</enumerationValue>
                        <enumerationValue key="ease_in_out_quad">Ease in out quardatic</enumerationValue>
                        <enumerationValue key="ease_in_quart">Ease in quartic</enumerationValue>
                        <enumerationValue key="ease_out_quart">Ease out quartic</enumerationValue>
                        <enumerationValue key="ease_in_out_quart">Ease in out quartic</enumerationValue>
                        <enumerationValue key="ease_in_quint">Ease in quintic</enumerationValue>
                        <enumerationValue key="ease_out_quint">Ease out quintic</enumerationValue>
                        <enumerationValue key="ease_in_out_quint">Ease in out quintic</enumerationValue>
                        <enumerationValue key="ease_in_sine">Ease in sine</enumerationValue>
                        <enumerationValue key="ease_out_sine">Ease out sine</enumerationValue>
                        <enumerationValue key="ease_in_out_sine">Ease in out sine</enumerationValue>
                        <enumerationValue key="ease_in_back">Ease in back</enumerationValue>
                        <enumerationValue key="ease_out_back">Ease out back</enumerationValue>
                        <enumerationValue key="ease_in_out_back">Ease in out back</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="direction" type="enumeration" defaultValue="normal">
                    <caption>Direction</caption>
                    <description/>
                    <enumerationValues>
                        <enumerationValue key="normal">Normal</enumerationValue>
                        <enumerationValue key="reverse">Reverse</enumerationValue>
                        <enumerationValue key="alternate">Alternate</enumerationValue>
                        <enumerationValue key="alternate_reverse">Alternate reverse</enumerationValue>
                    </enumerationValues>
                </property> 
            </propertyGroup>

            <propertyGroup caption="Timing">
                <property key="duration" type="integer" defaultValue="200">
                    <caption>Duration</caption>
                    <description/>
                </property>
                <property key="delay" type="integer" defaultValue="0">
                    <caption>Delay</caption>
                    <description/>
                </property>
                <property key="count" type="integer" defaultValue="1">
                    <caption>Count</caption>
                    <description>0 is infinite</description>
                </property>
                <property key="condition" type="expression" required="false" defaultValue="true">
                    <caption>Condition</caption>
                    <description/>
                    <returnType type="Boolean" />
                </property>
            </propertyGroup>

            <propertyGroup caption="Events">
                <property key="afterAnimationAction" type="action" required="false">
                    <caption>After animation</caption>
                    <description/>
                </property>
            </propertyGroup>
        </propertyGroup>
    </properties>
</widget>
