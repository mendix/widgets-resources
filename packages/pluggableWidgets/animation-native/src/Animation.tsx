import { Component, ReactNode, createElement } from "react";
import { flattenStyles } from "@native-mobile-resources/util-widgets";
import { Animation as AnimationType, View, Easing } from "react-native-animatable";

import { AnimationProps } from "../typings/AnimationProps";
import { defaultAnimationStyle, AnimationStyle } from "./ui/Styles";
import { executeAction } from "@widgets-resources/piw-utils";

export type Props = AnimationProps<AnimationStyle>;
type Direction = "normal" | "reverse" | "alternate" | "alternate-reverse";

export class Animation extends Component<Props> {
    private readonly animationEndHandle = this.onAnimationEnd.bind(this);
    private readonly styles = flattenStyles(defaultAnimationStyle, this.props.style);

    render(): ReactNode {
        const { count, duration, content, easing, delay, direction } = this.props;
        const easingValue = easing.replace(/_/g, "-") as Easing;
        const directionValue = direction.replace(/_/g, "-") as Direction;
        const countValue = count === 0 ? "infinite" : count;
        this.validateProps(this.props);

        return (
            <View
                testID={this.props.name}
                animation={this.getAnimation()}
                duration={duration}
                delay={delay}
                direction={directionValue}
                easing={easingValue}
                iterationCount={countValue}
                onAnimationEnd={this.animationEndHandle}
                style={this.styles.container}
                useNativeDriver
            >
                {content}
            </View>
        );
    }

    private validateProps(props: Props): void {
        const { afterAnimationAction, count } = props;
        if (afterAnimationAction && count === 0) {
            this.log("After animation action can not be triggered by infinite count");
        }
        const { animationType, animationIn, animationOut, animationAttention } = this.props;
        if (animationType === "in" && animationIn === "none") {
            this.log("No 'Entry animation' is selected for animation type 'Entry'");
        }
        if (animationType === "attention" && animationAttention === "none") {
            this.log("No 'Attention animation' is selected for animation type 'Attention'");
        }
        if (animationType === "out" && animationOut === "none") {
            this.log("No 'Exit animation' is selected for animation type 'Exit'");
        }
        if (animationType === "in" && (animationOut !== "none" || animationAttention !== "none")) {
            this.log(
                "The 'Attention' an 'Exit' animation is ignored and should be set to 'None' when effect 'type' 'Entry' is selected"
            );
        }
        if (animationType === "attention" && (animationOut !== "none" || animationIn !== "none")) {
            this.log(
                "The 'Entrance' and 'Exit' animation is ignored and should be set to 'None' when effect 'type' 'Attention' is selected"
            );
        }
        if (animationType === "out" && (animationIn !== "none" || animationAttention !== "none")) {
            this.log(
                "The 'Entry' and 'Attention' animation is ignored and should be set to 'None' when effect 'type' 'Exit' is selected"
            );
        }
    }

    private log(message: string): void {
        // eslint-disable-next-line no-console
        console.warn(message);
    }

    private getAnimation(): AnimationType | undefined {
        const { animationType, animationIn, animationOut, animationAttention, condition } = this.props;
        const animation =
            animationType === "in" ? animationIn : animationType === "out" ? animationOut : animationAttention;

        if (!condition || (condition.status === "available" && condition.value === true)) {
            return animation === "none" ? undefined : animation;
        }
        return undefined;
    }

    private onAnimationEnd(): void {
        executeAction(this.props.afterAnimationAction);
    }
}
