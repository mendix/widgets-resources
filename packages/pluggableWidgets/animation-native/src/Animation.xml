<?xml version="1.0" encoding="utf-8" ?>
<widget id="com.mendix.widget.native.animation.Animation" pluginWidget="true" needsEntityContext="true" offlineCapable="true"
        supportedPlatform="Native"
        xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../node_modules/mendix/custom_widget.xsd">
    <name>Animation</name>
    <description>Container for animating content</description>
    <icon>iVBORw0KGgoAAAANSUhEUgAAAEAAAABACAYAAACqaXHeAAAABGdBTUEAALGPC/xhBQAAA0BJREFUeAHtmLGPEkEUxmGJEi2EVgsbK0NCdZ2dJCZ2pwkQEqxOSYhWJmcsKY0kVhoSclcYMQTI6VWanMHwB1wp2EhzxVnQQIkG8Htk97I7MzvMLUdyeG8Twsyb730z89vZZYZQiC8mwASYABNgAkyACTABJsAEmAATYAJMgAkwASbABJgAE7g4BMLrONVCoXBpNBptTqfTzXA4vDGbzW7QPFA+RvnQsqz9WCy2X61W/y6a39oBSKfTDzDRMiZ6Szc5aPrQbLdarc9ana7xLNpyuVxiMpk8xmDuwe+m7XmEAR5EIpGder3eNemnVCpZ3W73FbTbJnqXppxIJF4if+qKnRRXtgJwpy5jkm/QUxGTt056dBXQToOqoP057tQfV5NUhN9rBE87ecenDP8XTsX97QsAgwpns1m6c1tISNhJXQx6t9Fo7OB75jZyl2nyqH/F5647ril/R9t9Pwi07NH+SZNv0vQQ/tLjoASQz+evj8fjD3BN+Ti3o9Hoo1qt9lvVnslk3gLcU1WbXwxA3zWbzWdiO73whsPhT/hpn3kxT6zDvx+Px2+LL0ZpadKdXzB58k6RhrRiR/TMI1YU4wb1op3rkdLbftnJkyF5kJfHHBUJAC17xP3uvDs/ZWvdsZD9wpN8PSJFBQO0KFdsop86MRa0rvKSBoqBbJl2oNIiRm/7QJcqF0t3I5CZIknlJQFAnvPCU1hIIZXW+amTxAYBKRdQ5pscg9yFEpWXCsBCo/9JoAJgtDGxIai0R0sAknKxbI+X8POkqrwkABDterI0FR/tN02Ktgl+B6IAy/ZQjAWtq7wkALTJQQdtg07attYjxUGENkm0wzvVRTm0NRaT6GAjxoLWVV4SAAxkRpscdKKDMN8IkVYcDKD8QKwixg3qFdW5gE516KdvkK+VkAd5iSJpI+MIsFwu7lbYgbDM97ochiLLTFKX2+v1Jji+fkkmk3tYTXTSu4bPFXyo/AtL8iOe+Sd4ZN6TFjHthfNFezAYXIXojlYoN86Pw51OR3pcSer7CMg+5yNCJ0PAW58/RFaB7Sz/ElvF+NiTCTABJsAEmAATYAJMgAkwASbABJgAE2ACTIAJMAEmwATOK4F/LkNXt8iCraQAAAAASUVORK5CYII=</icon>
    <properties>
        <propertyGroup caption="General">

            <propertyGroup caption="Content">
                <property key="content" type="widgets" required="true">
                    <caption>Content</caption>
                    <description/>
                </property>
            </propertyGroup>

            <propertyGroup caption="Effect">
                <property key="animationType" type="enumeration" defaultValue="in">
                    <caption>Type</caption>
                    <description>An entrance animation goes from invisible to visible. An attention animation stays visible and returns to its original state. An exit animation goes from visible to invisible.</description>
                    <enumerationValues>
                        <enumerationValue key="in">Entrance</enumerationValue>
                        <enumerationValue key="attention">Attention</enumerationValue>
                        <enumerationValue key="out">Exit</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="animationIn" type="enumeration" defaultValue="fadeIn">
                    <caption>Entry animation</caption>
                    <description/>
                    <enumerationValues>
                        <enumerationValue key="none">None</enumerationValue>
                        <enumerationValue key="bounceIn">Bounce</enumerationValue>
                        <enumerationValue key="bounceInDown">Bounce down</enumerationValue>
                        <enumerationValue key="bounceInUp">Bounce up</enumerationValue>
                        <enumerationValue key="bounceInLeft">Bounce left</enumerationValue>
                        <enumerationValue key="bounceInRight">Bounce right</enumerationValue>
                        <enumerationValue key="fadeIn">Fade</enumerationValue>
                        <enumerationValue key="fadeInDown">Fade down</enumerationValue>
                        <enumerationValue key="fadeInDownBig">Fade down big</enumerationValue>
                        <enumerationValue key="fadeInUp">Fade up</enumerationValue>
                        <enumerationValue key="fadeInUpBig">Fade up big</enumerationValue>
                        <enumerationValue key="fadeInLeft">Fade left</enumerationValue>
                        <enumerationValue key="fadeInLeftBig">Fade left big</enumerationValue>
                        <enumerationValue key="fadeInRight">Fade right</enumerationValue>
                        <enumerationValue key="fadeInRightBig">Fade right big</enumerationValue>
                        <enumerationValue key="slideInDown">Slide down</enumerationValue>
                        <enumerationValue key="slideInUp">Slide up</enumerationValue>
                        <enumerationValue key="slideInLeft">Slide left</enumerationValue>
                        <enumerationValue key="slideInRight">Slide right</enumerationValue>
                        <enumerationValue key="zoomIn">Zoom</enumerationValue>
                        <enumerationValue key="zoomInDown">Zoom down</enumerationValue>
                        <enumerationValue key="zoomInUp">Zoom up</enumerationValue>
                        <enumerationValue key="zoomInLeft">Zoom left</enumerationValue>
                        <enumerationValue key="zoomInRight">Zoom right</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="animationAttention" type="enumeration" defaultValue="none">
                    <caption>Attention animation</caption>
                    <description/>
                    <enumerationValues>
                        <enumerationValue key="none">None</enumerationValue>
                        <enumerationValue key="bounce">Bounce</enumerationValue>
                        <enumerationValue key="flash">Flash</enumerationValue>
                        <enumerationValue key="pulse">Pulse</enumerationValue>
                        <enumerationValue key="rotate">Rotate</enumerationValue>
                        <enumerationValue key="rubberBand">Rubber band</enumerationValue>
                        <enumerationValue key="shake">Shake</enumerationValue>
                        <enumerationValue key="swing">Swing</enumerationValue>
                        <enumerationValue key="tada">Tada</enumerationValue>
                        <enumerationValue key="wobble">Wobble</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="animationOut" type="enumeration" defaultValue="none">
                    <caption>Exit animation</caption>
                    <description/>
                    <enumerationValues>
                        <enumerationValue key="none">None</enumerationValue>
                        <enumerationValue key="bounceOut">Bounce</enumerationValue>
                        <enumerationValue key="bounceOutDown">Bounce down</enumerationValue>
                        <enumerationValue key="bounceOutUp">Bounce up</enumerationValue>
                        <enumerationValue key="bounceOutLeft">Bounce left</enumerationValue>
                        <enumerationValue key="bounceOutRight">Bounce right</enumerationValue>
                        <enumerationValue key="fadeOut">Fade</enumerationValue>
                        <enumerationValue key="fadeOutDown">Fade down</enumerationValue>
                        <enumerationValue key="fadeOutDownBig">Fade down big</enumerationValue>
                        <enumerationValue key="fadeOutUp">Fade up</enumerationValue>
                        <enumerationValue key="fadeOutUpBig">Fade up big</enumerationValue>
                        <enumerationValue key="fadeOutLeft">Fade left</enumerationValue>
                        <enumerationValue key="fadeOutLeftBig">Fade left big</enumerationValue>
                        <enumerationValue key="fadeOutRight">Fade right</enumerationValue>
                        <enumerationValue key="fadeOutRightBig">Fade right big</enumerationValue>
                        <enumerationValue key="slideOutDown">Slide down</enumerationValue>
                        <enumerationValue key="slideOutUp">Slide up</enumerationValue>
                        <enumerationValue key="slideOutLeft">Slide left</enumerationValue>
                        <enumerationValue key="slideOutRight">Slide right</enumerationValue>
                        <enumerationValue key="zoomOut">Zoom</enumerationValue>
                        <enumerationValue key="zoomOutDown">Zoom down</enumerationValue>
                        <enumerationValue key="zoomOutUp">Zoom up</enumerationValue>
                        <enumerationValue key="zoomOutLeft">Zoom left</enumerationValue>
                        <enumerationValue key="zoomOutRight">Zoom right</enumerationValue>
                    </enumerationValues>
                </property>
            </propertyGroup>

            <propertyGroup caption="Timing">
                <property key="duration" type="integer" defaultValue="200">
                    <caption>Duration (ms)</caption>
                    <description/>
                </property>
                <property key="delay" type="integer" defaultValue="0">
                    <caption>Delay (ms)</caption>
                    <description/>
                </property>
            </propertyGroup>

            <propertyGroup caption="Condition">
                <systemProperty key="Visibility"/>
                <property key="condition" type="expression" required="false">
                    <caption>Expression</caption>
                    <description>The animation will start after it becomes visible while also obeying its timing delay. For already-visible content, an expression can be used to start the animation.</description>
                    <returnType type="Boolean" />
                </property>
            </propertyGroup>
        </propertyGroup>

        <propertyGroup caption="Advanced">

            <propertyGroup caption="Animation">
                <property key="easing" type="enumeration" defaultValue="ease">
                    <caption>Easing</caption>
                    <description>Specifies an animation's rate of change over time.</description>
                    <enumerationValues>
                        <enumerationValue key="linear">Linear</enumerationValue>
                        <enumerationValue key="ease">Ease</enumerationValue>
                        <enumerationValue key="ease_in">Ease in</enumerationValue>
                        <enumerationValue key="ease_out">Ease out</enumerationValue>
                        <enumerationValue key="ease_in_out">Ease in out</enumerationValue>
                        <enumerationValue key="ease_in_cubic">Ease in cubic</enumerationValue>
                        <enumerationValue key="ease_out_cubic">Ease out cubic</enumerationValue>
                        <enumerationValue key="ease_in_out_cubic">Ease in out cubic</enumerationValue>
                        <enumerationValue key="ease_in_circ">Ease in circle</enumerationValue>
                        <enumerationValue key="ease_out_circ">Ease out circle</enumerationValue>
                        <enumerationValue key="ease_in_out_circ">Ease in out circle</enumerationValue>
                        <enumerationValue key="ease_in_expo">Ease in exponential</enumerationValue>
                        <enumerationValue key="ease_out_expo">Ease out exponential</enumerationValue>
                        <enumerationValue key="ease_in_out_expo">Ease in out exponential</enumerationValue>
                        <enumerationValue key="ease_in_quad">Ease in quardatic</enumerationValue>
                        <enumerationValue key="ease_out_quad">Ease out quardatic</enumerationValue>
                        <enumerationValue key="ease_in_out_quad">Ease in out quardatic</enumerationValue>
                        <enumerationValue key="ease_in_quart">Ease in quartic</enumerationValue>
                        <enumerationValue key="ease_out_quart">Ease out quartic</enumerationValue>
                        <enumerationValue key="ease_in_out_quart">Ease in out quartic</enumerationValue>
                        <enumerationValue key="ease_in_quint">Ease in quintic</enumerationValue>
                        <enumerationValue key="ease_out_quint">Ease out quintic</enumerationValue>
                        <enumerationValue key="ease_in_out_quint">Ease in out quintic</enumerationValue>
                        <enumerationValue key="ease_in_sine">Ease in sine</enumerationValue>
                        <enumerationValue key="ease_out_sine">Ease out sine</enumerationValue>
                        <enumerationValue key="ease_in_out_sine">Ease in out sine</enumerationValue>
                        <enumerationValue key="ease_in_back">Ease in back</enumerationValue>
                        <enumerationValue key="ease_out_back">Ease out back</enumerationValue>
                        <enumerationValue key="ease_in_out_back">Ease in out back</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="count" type="integer" defaultValue="1">
                    <caption>Cycles</caption>
                    <description>Specifies the number of times the animation is played (0 signifies an infinite loop).</description>
                </property>
                <property key="direction" type="enumeration" defaultValue="normal">
                    <caption>Direction</caption>
                    <description>Alternating direction will reverse the animation every even-numbered repeating cycle.</description>
                    <enumerationValues>
                        <enumerationValue key="normal">Normal</enumerationValue>
                        <enumerationValue key="alternate">Alternating</enumerationValue>
                    </enumerationValues>
                </property>
            </propertyGroup>

            <propertyGroup caption="Events">
                <property key="afterAnimationAction" type="action" required="false">
                    <caption>After animation</caption>
                    <description>Specifies an action triggered after the animation has completed (cannot be used with infinitely repeating animations).</description>
                </property>
            </propertyGroup>

            <propertyGroup caption="Common">
                <systemProperty key="Name"/>
            </propertyGroup>
        </propertyGroup>

    </properties>
</widget>
