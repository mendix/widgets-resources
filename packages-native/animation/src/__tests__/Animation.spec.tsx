import { createElement } from "react";
import { Text } from "react-native";
import { render } from "react-native-testing-library";

import { Animation, Props } from "../Animation";

jest.mock("Platform", () => ({
    OS: "ios",
    select: jest.fn(dict => dict.ios)
}));

const textContent = <Text>Hello</Text>;

const defaultProps: Props = {
    content: textContent,
    animationType: "in",
    animationIn: "none",
    animationAttention: "none",
    animationOut: "none",
    duration: 100,
    delay: 0,
    easing: "ease",
    count: 1,
    direction: "normal",
    style: []
};

describe("Animation", () => {
    it("renders", () => {
        const component = render(<Animation {...defaultProps} />);

        expect(component.toJSON()).toMatchSnapshot();
    });
});
