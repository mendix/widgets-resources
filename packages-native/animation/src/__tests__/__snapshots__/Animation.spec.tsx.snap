// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Animation renders 1`] = `
<View
  style={Object {}}
>
  <Text>
    Hello
  </Text>
</View>
`;
